gh-96461: clarify the meaning of the oparg for CACHE and COPY opcode (#96462)
python/cpython: Doc/library/dis.rst

**General instructions**

> In the following, TOS(i) refer to the *i*-th item on the stack which is index from 1.

Reviewer: ```suggestion
In the following, TOS(i) refers to the *i*-th item on the stack which is indexed from 1.
```

We also use as shorthand TOS = TOS(1) which is the top-of-stack.

> TOS1, TOS2, TOS3 are the second, third and fourth items on the stack, respectively.

Reviewer: It's a bit unfortunate that TOS1 = TOS(2),   TOS2 = TOS(3), etc. 

@markshannon - would you like to change this?

Author: This was another reason for my previous formulation that avoided ro shed too much light on this.

Reviewer 2: Generally we count from one for stack items, so I would like to change it. Ideally drop `TOS1` etc, all together.
The current explanation is not explicit about what is popped or not, so the shorthand is not helpful.

.. opcode:: POP_TOP

   Removes the top-of-stack item.::

      STACK.pop()

.. opcode:: COPY (i)

>    Push the i-th to the top of the stack without removing it from its original

Reviewer: ```suggestion
   Push the i-th item to the top of the stack without removing it from its original
```

   location.::

>       STACK.append(STACK[-i]

Reviewer: ```suggestion
      STACK.append(STACK[-i])
```

>       STACK.append(STACK[-i])

Reviewer 3: Add that `i` should not be zero.

   Push the *i*-th item to the top of the stack. The item is not removed from its
   original location.

   The stack is indexed from 1, so COPY 1 will copy TOS.

   .. versionadded:: 3.11

.. opcode:: SWAP (i)

   Swap TOS with the item at position *i*.

   The stack is indexed from 1, so SWAP 2 will swap the two element at the top
   of the stack.

> 

Reviewer: Perhaps the explanation of TOS etc under the "**Binary and in-place operations**" heading should be moved up?

Author: I considered doing so. But since it was a bigger change I though I would wait to see what reviewers thought. 

I will move it as soon as I get a chance since we agree it makes sense.

.. opcode:: COPY (i)

>    Push TOS(i) to the top of the stack. The item is not removed from its

Reviewer 2: Maybe "Push a copy of `i`th item (counting from 1) to the top of the stack ..."

   original location.

   .. versionadded:: 3.11

.. opcode:: SWAP (i)

>    Swap TOS with TOS(i).

Reviewer 2: Maybe "Swap the first and `i`th item on the stack (counting from 1).

   Push the *i*-th item to the top of the stack. The item is not removed from its
   original location.

>    The stack is indexed from 1, so COPY 1 will copy TOS.

Reviewer: Maybe instead of adding this, we could define ``TOS(i)`` at the top along with TOS1, TOS2 etc, and then reword the definition here to use those.

Author: I tried something along those lines in my last commit let me know if you like it better this way.

.. opcode:: UNARY_POSITIVE

   Implements ``STACK[-1] = +STACK[-1]``.

.. opcode:: UNARY_NEGATIVE

   Implements ``STACK[-1] = -STACK[-1]``.

.. opcode:: UNARY_NOT

   Implements ``STACK[-1] = not STACK[-1]``.

.. opcode:: UNARY_INVERT

   Implements ``STACK[-1] = ~STACK[-1]``.

.. opcode:: GET_ITER

   Implements ``STACK[-1] = iter(STACK[-1])``.

.. opcode:: GET_YIELD_FROM_ITER

   If ``STACK[-1]`` is a :term:`generator iterator` or :term:`coroutine` object
   it is left as is.  Otherwise, implements ``STACK[-1] = iter(STACK[-1])``.

   .. versionadded:: 3.5

**Binary and in-place operations**

Binary operations remove the top two items from the stack (``STACK[-1]`` and
``STACK[-2]``). They perform the operation, then put the result back on the stack.

In-place operations are like binary operations, but the operation is done in-place
when ``STACK[-2]`` supports it, and the resulting ``STACK[-1]`` may be (but does
not have to be) the original ``STACK[-2]``.

.. opcode:: BINARY_OP (op)

   Implements the binary and in-place operators (depending on the value of
   *op*).

>    ``STACK[-1] =  op STACK[-1]``.

Reviewer 2: You are missing the left operand.
`STACK[-1] = STACK[-2] op STACK.pop()`
or
```
rhs = STACK.pop()
lhs = STACK.pop()
STACK.push(lhs op rhs)
```

   .. versionadded:: 3.11

.. opcode:: BINARY_SUBSCR

>    Implements ``STACK[-1] = STACK[-2][STACK[-1]]``.

Reviewer 2: This is incorrect, although no more than the old version.
See my comments on `BINARY_OP`

.. opcode:: BUILD_TUPLE (count)

   Creates a tuple consuming *count* items from the stack, and pushes the
   resulting tuple onto the stack.::

>       STACK.append(tuple(STACK[-count:]))

Reviewer 2: This isn't quite correct. The colon should be before the `-count` and the values need to be popped from the stack.
Also state that `count` should not be zero (it can be, but the compiler emits `LOAD_CONST` in that case, so we might change that it the future.
```
    assert(count > 0)
    STACK, values = STACK[-count:], STACK[:-count]
    STACK.append(tuple(values))
```

Author: I am confused. `BUILD_TUPLE` consume the items on top of the stack so if count is 2 that would be `STACK[-2],STACK[-1]  == STACK[-2:]` so shouldn't we write:
```
assert count > 0
STACK, values = STACK[:-count], STACK[-count:]
STACK.append(tuple(values))
```

Reviewer 2: Yes, you're right.
